sql: support backslash_quote session setting

ZhouXing19 opened the pull request:
This setting is based on a previous issue: cockroachdb#45774. This commit set a placeholder for PostgreSQL configuration setting called [backslash_quote](https://www.postgresql.org/docs/12/runtime-config-compatible.html#GUC-BACKSLASH-QUOTE), and ensured it is compatible with ORMS. Currently, we only allow the value to be `backslash_quote`.  If `client_encoding` is updated to allow encodings other than UTF8, then the default value of `backslash_quote` should be changed to `on`.

When users try setting the backslash_quote session variable, CRDB will let it pass, but not making any changes. We leave the functionality implementation to future works.

Details:
- Added `backslash_quote` at varGen at `sql/vars.go` without implementing the true functionality

- Tested it with `subtest backslash_quote_test` at `pkg/sql/logictest/testdata/logic_test/set`

- Removed `backslash_quote` from `unsupported_vars.go`

Passed local test:
<img src="https://i.imgur.com/crZFk9w.png">

Updates:

- nil: updated url for backslash_quote, and removed extra spaces

- Deleted `backslash_quote` from `unsupported_vars.go`

- Ran `make testlogic FILES=information_schema TESTFLAGS=-rewrite`

Release note (sql change): Added session variable backslash_quote for compatibility, setting it does a no-op, and only `safe_encoding` is supported.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67343)
<!-- Reviewable:end -->

RichardJCai commented on pkg/sql/vars.go:
```diff
@@ -831,8 +831,11 @@ var varGen = map[string]sessionVar{
 	// See https://www.postgresql.org/docs/10/static/runtime-config-client.html#GUC-LOC-TIMEOUT
 	`lock_timeout`: makeCompatIntVar(`lock_timeout`, 0),
 
+	// See https://www.postgresql.org/docs/9.1/runtime-config-compatible.html
```
nit: use the latest version of the postgresql docs.
https://www.postgresql.org/docs/13/runtime-config-compatible.html

RichardJCai commented on pkg/sql/unsupported_vars.go:
```diff
@@ -64,7 +64,7 @@ var UnsupportedVars = func(ss ...string) map[string]struct{} {
 	"array_nulls",
 	"backend_flush_after",
-	"backslash_quote",
+	//"backslash_quote",
```
I think we can just remove this line

RichardJCai commented on .idea/vcs.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Remove these.xml files, you should add .idea to a global gitignore.

RichardJCai commented on pkg/sql/logictest/testdata/logic_test/set:
```diff
@@ -476,3 +476,20 @@ SET standard_conforming_strings='true'
 
 statement ok
 SET standard_conforming_strings='on'
+
```
nit: extra space

RichardJCai commented on pkg/sql/logictest/testdata/logic_test/set:
```diff
@@ -476,3 +476,20 @@ SET standard_conforming_strings='true'
+
+statement error invalid value for parameter "backslash_quote"
+SET backslash_quote = '123';
+
```
nit: extra space

RichardJCai reviewed (commented):
Great job on your first PR. 🥳 Looks really good, just some nits.

Please update the commit message with 
```
Release note (sql change): Added session variable backslash_quote for compatibility, setting it does a no-op.
```

See other PR commits for examples, there is also an wiki page on how we use release notes, I can send that to you.

RichardJCai commented:
Also there's some logictests that you'll have to update, you can look at them on the GitHub CI.
![image](https://user-images.githubusercontent.com/25163644/124852227-00f11f00-df72-11eb-9614-0c24aea1471f.png)

To update these, you can do `make testlogic FILES="variables show_source pg_catalog" TESTFLAGS=-rewrite` 
This will update the output of the logic tests with the new variable you added.

rewrite testflag is useful to keep in mind but be careful when using it. Let me know if you have any questions.

ZhouXing19 commented:
bors r+

RichardJCai commented:
bors r-

RichardJCai commented on .gitignore:
```diff
@@ -39,3 +39,6 @@ build/Railroad.jar
 
 # Per-user .bazelrc
 /.bazelrc.user
+
```
Please don't commit this file

RichardJCai commented on pkg/sql/vars.go:
```diff
@@ -831,8 +831,11 @@ var varGen = map[string]sessionVar{
+
 	// Supported for PG compatibility only.
-	// See https://www.postgresql.org/docs/10/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
+	// See https://www.postgresql.org/docs/10/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIER
```
Did you mean to remove the S here?

RichardJCai reviewed (commented):
Please wait until the PR is "Approved" before running bors

ZhouXing19 committed:
sql: support backslash_quote session setting

This setting is based on an previous issue: #45774. This commit set a placeholder for PostgreSQL configuration setting called (backslash_quote)[https://www.postgresql.org/docs/13/runtime-config-compatible.html#GUC-BACKSLASH-QUOTE], and ensured it is compatible with ORMS. Currently, we only allow the value to be `backslash_quote`.  If `client_encoding` is updated to allow encodings other than UTF8, then the default value of `backslash_quote` should be changed to `on`.

When users try setting the backslash_quote session variable, CRDB will let it pass, but not making any changes. We leave the functionality implementation to future works.

Details:
- Added `backslash_quote` at varGen at `sql/vars.go` without implementing the true functionality

- Tested it with `subtest backslash_quote_test` at `pkg/sql/logictest/testdata/logic_test/set`

- Removed `backslash_quote` from `unsupported_vars.go`

Updates:

- nil: updated url for backslash_quote, and removed extra spaces

- Deleted `backslash_quote` from `unsupported_vars.go`

- Added `./.idea` at `.gitignore`

- Ran `make testlogic FILES=information_schema TESTFLAGS=-rewrite`

Release note (sql change): Added session variable backslash_quote for compatibility, setting it does a no-op, and only safe_encoding is supported.

RichardJCai reviewed (approved):
Awesome work on your first PR. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  once CI passes! :tada: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67343#-:-Me6CdgIAzdqXCwpkUl7:b-oj23bs)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss, @RichardJCai, and @ZhouXing19)



<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/vars.go:
```diff
@@ -831,6 +831,9 @@ var varGen = map[string]sessionVar{
 	`lock_timeout`: makeCompatIntVar(`lock_timeout`, 0),
 
+	// See https://www.postgresql.org/docs/13/runtime-config-compatible.html
+	`backslash_quote`: makeCompatStringVar(`backslash_quote`, `safe_encoding`, `on`, `off`, `safe_encoding`),
```
for this PR, I think we should only allow the values that we actually support.

I think it would be misleading for users if we allow them to set this to a value, but we don't actually implement the behavior described in the postgres docs.

rafiss replied:
so to be a little more specific, I think this line should be:
```
makeCompatStringVar(`backslash_quote`, `safe_encoding`),
```

It also needs a comment saying that if we allow other values for `client_encoding` that are not UTF-8, then we need to update this to either implement `safe_encoding` fully, or change the default value to `on`.


To fully implement `backslash_quote=off`, we would need to update the code here so that it uses this session variable to determine the value of `allowEscape`: https://github.com/cockroachdb/cockroach/blob/36ed61e139271756805545ae5939c40fb32e7e45/pkg/sql/parser/scan.go#L155-L163

But for this PR, I don't think we should fully implement the setting. @RichardJCai @ZhouXing19 -- what do you think?

RichardJCai replied:
Yeah I don't think we should fully implement it, let's not support `off` for now.

ZhouXing19 replied:
@rafiss @RichardJCai Cool, have changed the `extraAllowed` to an empty set, and modified the tests at `logic_test\set`.

rafiss reviewed (changes requested):
this is looking good! i just have a comment about which values we should allow though

RichardJCai commented:
Reminder to update the release note saying only `safe_encoding` is supported.

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67343)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss and @RichardJCai)

___
*[pkg/sql/unsupported_vars.go, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67343#-Me2PqqS2ckMaXY0KViH-r1-67:-Me7cvAcD0ZsED661sGN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80d8d09ff1802c3986175e4a1d1b9c4bf2b5a00b/pkg/sql/unsupported_vars.go#L67)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I think we can just remove this line
</blockquote></details>

Done.
___
*[pkg/sql/vars.go, line 838 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67343#-Me2PqqS2ckMaXY0KViI-r2-838:-Me7cyzOCHwfxaF__OtQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcbc4870aa7aecb693ecc0df4155861317e3e763/pkg/sql/vars.go#L838)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

@RichardJCai Oops, I deleted it by accident. Will fix it!
</blockquote></details>

Done.
___
*[pkg/sql/vars.go, line 835 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67343#-Me2PqqS2ckMaXY0KViI-r3-835:-Me7cxoZBinBIDhtUZXH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2351877accc846ea9ececb9c2a54dde26ead4d4/pkg/sql/vars.go#L835)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

@rafiss @RichardJCai Cool, have changed the `extraAllowed` to an empty set, and modified the tests at `logic_test\set` and commit/PR comments.
</blockquote></details>

Done.
___
*[.gitignore, line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67343#-Me3Ff5hmWk94ywqm1Lc-r2-42:-Me7d02498UjWzd8fAd8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcbc4870aa7aecb693ecc0df4155861317e3e763/.gitignore#L42)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Please don't commit this file
</blockquote></details>

Done.
___
*[.idea/vcs.xml, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67343#-Me2PqqS2ckMaXY0KViN-r1-1:-Me7d10PAcSO4BBeLL4v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80d8d09ff1802c3986175e4a1d1b9c4bf2b5a00b/.idea/vcs.xml#L1)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Remove these.xml files, you should add .idea to a global gitignore.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
this looks good to me!

i'm pretty sure this is failing because of some unrelated flaky tests. somebody else skipped them, so if you run `git pull --rebase origin master` (or perhaps another name like `upstream` if you named the remote something else other than `origin`), then you will get those changes and you can resubmit

ZhouXing19 commented:
bors r+
